### PR TITLE
Web DV Client SDK

### LINKED ISSUE #3: Create DataVaultProvider Web Client SDK
- Receives:
    - `web3Provider`/`signer` function and a did. Will be used to authenticate
    - Optional: storage get and set functions. Will be used to save the access token so it can be reused. Default: [window.localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage). This is an option for other clients, like mobile applications.
- Implements `@rsksmart/express-did-auth` authentication protocol - [`RIFID-450`](https://rsklabs.atlassian.net/browse/RIFID-450)
  - The login will be done in the first interaction with the DV API and the token will be saved in the provided storage so it can be reused
- Speaks with DV Service API
- Encrypts/decrypts content prior to put or after retrieving it from the DV Service
- WebJS compatible

```typescript=
type Key = string
type ContentIdentifier = string
type Content = string // if it is a JSON, must be stringified

type PutFilePayload = { key: Key, content: Content }
type PutFileResponse = { id: ContentIdentifier }
type GetFilePayload = { key: Key }
type GetFileResponse = { id: ContentIdentifier, content: Content }
type DeleteFilePayload = { key: Key, id: Identifier }
type DeleteFileResponse = {}
type SwapFilePayload = { key: Key, id: ContentIdentifier, content: Content }
type SwapFileResponse = { id: ContentIdentifier }

interface DataVaultWebClient {
  put(payload: PutFilePayload): PutFileResponse
  put(payload: PutFilePayload[]): PutFileResponse[]
  get(payload: GetFilePayload): GetFileResponse[]
  get(payload: GetFilePayload[]): GetFileResponse[]
  delete(payload: DeleteFilePayload): DeleteFileResponse
  delete(payload: DeleteFilePayload[]): DeleteFileResponse[]
  swap(payload: SwapFilePayload): SwapFileResponse
  swap(payload: SwapFilePayload[]): SwapFileResponse[]
}
```

### PR DESCRIPTION
**Pull request by javiesses**
Closes #3 

- [x] Deleted old version
- [x] Get content
- [x] Create content
- [x] Login process
- [x] Tests for every written method
- [x] Fixed typeorm version across the entire repo
- [x] README brief
- [x] Swap content
- [x] Delete content
- [x] Refresh access token when necessary
- [x] Login again when refresh token is expired
- [x] Refactor internal methods

Open work:
- Perform bulk CRUD operations - #13 
- Encryption box implementation - #14

### CODE REVIEW

**Review by ilanolkies — COMMENTED** [review]
First of all,
> Overrides to perform operation with multiple values (get content from multiple keys for instance)
do we want to do this or use different method names?

**Review by ilanolkies — CHANGES_REQUESTED** [review]
Some minor changes we talked about:
- Activate `test:ci` script to enable Circle CI before merging
- DID Auth token management is white box. That lead into two conclusions: analyse wether we need to test this or not, and remove access token related methods from client interface.

File: modules/ipfs-cpinner-client/src/auth-manager.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  const login = async (): Promise<LoginResponse> => {
+    const { did, signer, serviceUrl } = config
+    if (!did) throw new Error(NO_DID)
+    if (!signer) throw new Error(NO_SIGNER)
+
+    const tokens = await getChallenge()
+      .then(signChallenge)
+      .then(signature => axios.post(`${serviceUrl}/auth`, { response: signature }))
+      .then(res => res.status === 200 && res.data)
+      .catch(console.log)
```
Comment (ilanolkies) [inline modules/ipfs-cpinner-client/src/auth-manager.ts:16]: This hides the error

File: modules/ipfs-cpinner-client/src/constants.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export const ACCESS_TOKEN_KEY = 'accessToken'
```
Comment (ilanolkies) [inline modules/ipfs-cpinner-client/src/constants.ts:1]: I would use a data vault specific mey. Like `DATA_VAULT_QCCESS_TOKEN`

File: modules/ipfs-cpinner-client/src/auth-manager.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  const getChallenge = async (): Promise<string> => {
+    const { did, serviceUrl } = config
+    if (!did) throw new Error(NO_DID)
+
+    return axios.get(`${serviceUrl}/request-auth/${did}`)
+      .then(res => res.status === 200 && !!res.data && res.data.challenge)
+      .catch(console.log)
+  }
+
+  const signChallenge = async (challenge: string): Promise<string> => {
```
Comment (ilanolkies) [inline modules/ipfs-cpinner-client/src/auth-manager.ts:57]: This is what we need to change for new DI Auth integration. I identify it now to link it in an issue.

File: modules/ipfs-cpinner-client/src/index.ts [hunk 4]
```diff
@@ -1,98 +1,89 @@
...
+  ClientKeyValueStorage, CreateContentPayload, CreateContentResponse,
+  DeleteTokenPayload, GetContentPayload, Config,
+  SwapContentPayload, SwapContentResponse
+} from './types'
 
-interface ICentralizedIPFSPinnerClient {
-  put: (key: string, content: string) => Promise<string>
-  get: (key: string) => Promise<{ cid: string, content: Buffer }[] | { cid: string }[]>
-  delete: (key: string, cid: string) => Promise<string>
+const ClientKeyValueStorageFactory = {
```
Comment (ilanolkies) [inline modules/ipfs-cpinner-client/src/index.ts:12]: I would move it to a new file and remove it from the exports. User then point it by location

Comment (ilanolkies) [reply modules/ipfs-cpinner-client/src/index.ts:12]: In the future we can even move it out and configure it as a pluggable, making mobile apps and such not dependent on this

File: modules/ipfs-cpinner-client/README.md [hunk 5]
```diff
@@ -1,3 +1,180 @@
...
+
+## Features
+
+- Manage authentication according to the [DID Auth protocol](https://rsksmart.github.io/rif-identity-docs/ssi/libraries/express-did-auth)
+
+- CRUD operations over the RIF Data Vault
+
+- Stores the authentication credentials in the given storage 
+
+## Usage
```
Comment (ilanolkies) [inline modules/ipfs-cpinner-client/README.md:24]: I would move all this section directly to docs. It is very big for a readme.

```suggestion
## Usage

[Read the docs](https://rsksmart.github.io/rif-identity-docs/data-vault/cpinner/cpinner-client)

```

I will then review and iterate docs over that PR

**Review by ilanolkies — APPROVED** [review]
Fantastic work!

**Review by ilanolkies — CHANGES_REQUESTED** [review]
I missed the option. I now put request changes for stuff described above.

File: modules/ipfs-cpinner-client/src/auth-manager.ts (hunk 3 above)
Comment (javiesses) [reply modules/ipfs-cpinner-client/src/auth-manager.ts:57]: #15

File: modules/ipfs-cpinner-client/src/index.ts (hunk 4 above)
Comment (ilanolkies) [reply modules/ipfs-cpinner-client/src/index.ts:12]: Opened https://github.com/rsksmart/rif-data-vault/issues/16

Comment (javiesses) [reply modules/ipfs-cpinner-client/src/index.ts:12]: I'm just exporting the interface I want to be implemented when sending a custom storage, I think we should keep it there

**Review by ilanolkies — APPROVED** [review]
Fantastic

**Pull request merged**